experimental/QuadFormAndIsom/src/enumeration.jl

```diff
@@ -115,34 +122,58 @@ julia> is_admissible_triple(genus(F), genus(C), genus(Lf), 5)
 true
 ```
 """
-function is_admissible_triple(A::ZZGenus, B::ZZGenus, C::ZZGenus, p::IntegerUnion)
-  zg = genus(integer_lattice(; gram=matrix(QQ, 0, 0, [])))
+function is_admissible_triple(
+    A::ZZGenus,
+    B::ZZGenus,
+    C::ZZGenus,
+    p::Int,
+  )
   AperpB = direct_sum(A, B)
+
+  # Signatures conditions
+  # A+B and C must have the same signatures
   (signature_tuple(AperpB) == signature_tuple(C)) || (return false)
-  if ((A == zg) && (B == C)) || ((B == zg) && (A == C))
+  if ((rank(A) == 0) && (B == C)) || ((rank(B) == 0) && (A == C))
     # C can be always glued with the empty genus to obtain C
     return true
-  elseif (A == zg) || (B == zg)
+  elseif (rank(A) == 0) || (rank(B) == 0)
     # If A or B is empty but the other is not C, then there is no gluing
     return false
   end
 
-  @req is_divisible_by(rank(B), p-1) "p-1 must divide the rank of B"
-
+  # Rank condition
+  # The rank of B should be divisible by p-1
+  !is_divisible_by(rank(B), p-1) && return false
+
+  # Gluing condition
+  # The determinant of C should divide the one of A+B, and the quotient should
+  # be an even power of the prime p
+  _q, _r = divrem(numerator(det(AperpB)), numerator(det(C)))
+  !iszero(_r) && return false
+  ok, _g, _p = is_prime_power_with_data(_q)
+  !ok && return false
+  _p != p && return false
+  g, r = divrem(_g, 2)
+  !iszero(r) && return false
+
```

Comment:
Looks more complicated but eventually that is better. Instead of running over a finite set of value and do a lot of multiplication, we make sure that the determinants satisfy the wanted proprety ($det(A)*det(B) = p^{2g}det(C)$) and then we make sure that $g$ is small enough.